### PR TITLE
TS: Added additional type information toArray( ) to support valid usage.

### LINKED ISSUE #19303: TS: Vector3 interface not descriptive enough to support this valid usage
##### Description of the problem

The following code works, but TypeScript does not compile with the present Vector3.d.ts

```
  // vector between center of head and center of ear
  const earVector = new THREE.Vector3(1.44, 0.6, -0.3);

  const leftEar = new THREE.SphereGeometry(1, 32, 32).applyMatrix4(new THREE.Matrix4().makeScale(0.4, 0.4, 0.25));
  const rightEar = leftEar.clone();

  leftEar.translate(...earVector.toArray());
  earVector.x = -earVector.x;
  rightEar.translate(...earVector.toArray());
```
BUG (compile error):
`Expected 3 arguments, but got 0 or more. ts(2556)`

EXPECTED:
Given that the code behaves as expected, I would expect it to compile.
(I would also expect the aesthetics of this use of the spread operator to divide opinion)

##### Three.js version

- [ ] Dev
- [x] r116
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

##### Hardware Requirements (graphics card, VR Device, ...)
N/A

### PR DESCRIPTION
**Pull request by Antony74**
Added additional type information to Vector and Matrix toArray() methods.  This allows code involving the spread operator for function parameters (which works as expected) to compile in TypeScript.

Related issues:

Fixed #19303.

**Description**

This is the solution - defining tuples for brevity - was proposed by others and well recieved in my original bug report, so I thought I should write these definintions.

### CODE REVIEW

File: src/math/Matrix3.d.ts
Comment (mrdoob) [inline src/math/Matrix3.d.ts:134]: Why `offset?: 0`?

Comment (Antony74) [reply src/math/Matrix3.d.ts:134]: That's a good question, thanks.  The special case I've inserted does indeed depend upon the offset being zero.  If the offset changes, then the length of the array returned changes, so it is no longer a Matrix3tuple, or whatever, so TypeScript should use the existing more general definition of toArray in those circumstances.

File: src/math/Vector2.d.ts
Comment (donmccurdy) [inline src/math/Vector2.d.ts:432]: I believe this valid usage will cause a compilation error...

```js
const tmp = [];
vector.toArray(tmp);
```

... because the argument is not of type Vector2tuple.

File: src/math/Matrix3.d.ts
Comment (donmccurdy) [inline src/math/Matrix3.d.ts:8]: Consider using shorter gl-matrix style names here, like `mat4`: http://glmatrix.net/docs/.

File: src/math/Matrix3.d.ts
Comment (donmccurdy) [reply src/math/Matrix3.d.ts:134]: I believe the argument should remain a `number[]` here — it could have any length — but the return type could be a tuple.

File: src/math/Vector2.d.ts
Comment (Antony74) [reply src/math/Vector2.d.ts:432]: Thanks for providing the test case.  It does compile for me.  If I right-click on your call to toArray in Visual Studio Code and choose "Go to definition..." from the context menu, it takes me to the existing definition of toArray, not the special case that I have inserted.  The compiler seems to be able to match the correct definition as required.  I imagine you'll want to play with it yourself to satisfy yourself of this and try and find some means of breaking the definition that I have provided.

File: src/math/Matrix3.d.ts
Comment (Antony74) [reply src/math/Matrix3.d.ts:134]: That's a good thought, but I can change the return type by changing the length of the array in that parameter.

image

Comment (donmccurdy) [reply src/math/Matrix3.d.ts:134]: Oh, the actual return type of the method can actually be `number[]` then, yes. You could handle this with generics:

```typescript
toArray<T = number[] | vec2>(target: T): T;
```

That said, I'm not sure generics would actually make type-checking work any better here, compared to just letting the user cast the result:

```typescript
const result = vector.toArray([]) as vec2;
```

File: src/math/Matrix3.d.ts
Comment (Antony74) [reply src/math/Matrix3.d.ts:8]: That's cool, yes you posted those definitions to my bug report.  That was very helpful, I used that as my starting point.

I'm happy to go with whatever the prefered naming conventions are around here, but first I should explain why according to my own taste I extended those names, so that `mat4` became `Matrix4tuple`

- To disambiguate type-names from variable-names I uppercase the first letter of a type, and lowercase the first letter of a variable.  [I think this is fairly standard](https://basarat.gitbook.io/typescript/styleguide#type).
- I added the word `tuple` to make it clear that this type is a tuple, and thus help to disambiguate it from the Matrix4 type. 
- `mat` became `matrix` to make it clear to those not familiar with mathematical shorthand that this isn't something to wipe your feet on (or a material or mesh with a mat-like quality)

Having clarified that, I'll naturally be more than happy to restore the original type names if those are prefered.

File: src/math/Matrix3.d.ts
Comment (Antony74) [reply src/math/Matrix3.d.ts:134]: Yes, casting the result is a valid workaround for the original bug anyway.

That's clever to use generics to try and combine the two alternative definitions - the original more general one, and my more specific one - into a single definition.  It doesn't work because of how the value given for the offset parameter changes the return type.  Code utilising generics is sometimes very powerful, but it is harder to read.  Remember that the various definitions of `toArray()` do appear in the context-senstive help while you are entering its parameters via an IDE such as Visual Studio Code.

Comment (donmccurdy) [reply src/math/Matrix3.d.ts:134]: So with two function overloads available, the compiler chooses?

```typescript
// loose definition
const tmp = [];
vector.toArray(tmp); // → number[], matches loose definition
vector.toArray(tmp as vec3, 100); // number[], matches loose definition

// strict definition
const tmp = [] as vec3;
vector.toArray(tmp); // → vec3, matches subclass definition
vector.toArray(tmp, 0); // → vec3, matches subclass definition
```

I think I see now. 👍

File: src/math/Matrix3.d.ts
Comment (donmccurdy) [reply src/math/Matrix3.d.ts:8]: The names used by gl-matrix are also used by the WebGL API itself, e.g. https://thebookofshaders.com/glossary/?search=mat4, so to me it's more of a WebGL idiom here. But if we did want a more explicit name I'd suggest `Matrix4Tuple`, `ArrayMatrix4`, or `Matrix4Array`. Since we shouldn't need to export this type definition, maybe it doesn't make too much difference.

Comment (Antony74) [reply src/math/Matrix3.d.ts:8]: That's cool!  Someone will have to decide about short names vs. more explicit names, though they may want to let the dust settle here first.

But for explicit names, I will at minimum need to upper-case the 'T' in 'tuple' to convert my suggestion of `Matrix4tuple` into your `Matrix4Tuple`, thus correcting my SnakeCase.

Comment (donmccurdy) [reply src/math/Matrix3.d.ts:8]: Ok – Yeah I'm fine with any of the names here really, since it's not exported. If others have feedback feel free to comment, but I don't mean to block the PR on a change to the names.

Comment (Antony74) [reply src/math/Matrix3.d.ts:8]: Thanks :-)  SnakeCase fixed as promised.

File: src/math/Matrix3.d.ts
Comment (Antony74) [reply src/math/Matrix3.d.ts]: Let me know if there's anything further I can be of assistance with.  I believe I have answered all your questions.  I appreciate it's difficult to find the time to sit down and play with my new definition, and I'm not trying to hurry you.  It's entirely up to you guys what you decide and when.

The idea here is simply to provide more detailed type information for the most common usages of `toArray`.  Specifically, in order to use the spread operator (`...`) with methods such as `translate` which require a specific number of arguments the compiler needs to know, not just that `toArray` returns an array, but also that the length of that returned array matches the number of arguments.

As @donmccurdy has already figured out, this change is non-breaking because the compiler is happy so long as it can find one definition of `toArray` that matches, so adding an additionally definition shouldn't be able to take away from what's already there.  Of course that's not enough on it's own, you also need to check that the definition is correct, because if it transpired it wasn't and you had to correct it, that would almost certainly break code compiled against the definitions I have added here.  I don't think it's too hard to see that the new definitions simply describe a particular case of what `toArray` already does, and I wish you well with coming to your decision about this.

**Review by donmccurdy — APPROVED** [review]

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**
